[12.0][IMP] account_invoice_inter_company: onchange_helper dep

chienandalu opened the pull request:
Removes `onchange_helper` dependency and uses fancy `Form` class for the equivalent one (and even more complete and bullet-proof).

cc @Tecnativa TT21089

chienandalu committed:
[FIX] account_invoice_inter_company: onchange_helper dep

pedrobaeza commented on account_invoice_inter_company/models/account_invoice.py:
```diff
@@ -161,32 +162,26 @@ def _prepare_invoice_data(self, dest_company):
+            default_type=dest_inv_type))
+        dest_invoice_data.company_id = dest_company
+        dest_invoice_data.partner_id = self.company_id.partner_id
         return {
```
Do:
```python
vals = dest_invoice_data._values_to_save(all_fields=True)
vals.update({
    'name': self.name,
    'origin': ...
```
The idea of this is to recover all the possible extra values that other modules can inject in the invoice. You can remove also a lot of hardcoded values (or event better, assign them to the form instance instead of later, for also have the corresponding onchanges played).

Do the same on the invoice lines.

chienandalu replied:
Done in invoice, but in invoice lines I'm having this issue following that method:

current `vals`:
```python
{ 
    'account_analytic_id':	False,
    'account_id':	False,
    'analytic_tag_ids':	[(
        6, 
        False, 
        []
        )],
    'company_id':	False,
    'currency_id':	False,
    'discount':	0,
    'display_type':	False,
    'invoice_line_tax_ids':	[(
        6, 
        False, 
        [2]
        )],
    'invoice_type':	False,
    'name':	False,
    'partner_id':	False,
    'price_unit':	1.0,
    'product_id':	66,
    'quantity':	1.0,
    'sequence':	10,
    'uom_id':	6,
}
```

`vals` generated with `vals = invoice_line_dest._values_to_save(all_fields=True)`:
```python
{
    'account_analytic_id':	account.analytic.account(),
    'account_id':	237,
    'auto_invoice_line_id':	54,
    'company_id':	44,
    'discount':	0.0,
    'invoice_id':	63,
    'invoice_line_tax_ids':	[2],
    'name':	'Service Multi Company',
    'partner_id':	167,
    'price_unit':	450.0,
    'product_id':	66,
    'quantity':	1.0,
    'sequence':	10,
}
``

pedrobaeza commented on account_invoice_inter_company/models/account_invoice.py:
```diff
@@ -161,36 +162,28 @@ def _prepare_invoice_data(self, dest_company):
             'reference': self.number,
-            'account_id': dest_partner_data.get('account_id', False),
-            'partner_id': self.company_id.partner_id.id,
             'journal_id': dest_journal.id,
```
Do these assignations on the form, not later, as they can trigger things.

chienandalu commented on account_invoice_inter_company/views/account_invoice_views.xml:
```diff
@@ -0,0 +1,18 @@
+        <field name="mode">primary</field>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='invoice_line_ids']/tree" position="inside">
+                <field name="company_id" invisible="1" readonly="0"/>
```
Already in the main view. Could it be causing the error in travis?
https://github.com/odoo/odoo/blob/12.0/addons/account/views/account_invoice_view.xml#L492

chienandalu commented on account_invoice_inter_company/views/account_invoice_views.xml:
```diff
@@ -0,0 +1,18 @@
+                <field name="company_id" invisible="1" readonly="0"/>
+                <field name="partner_id" invisible="1" readonly="0"/>
+                <field name="invoice_type" invisible="1" readonly="0"/>
+                <field name="currency_id" invisible="1" readonly="0"/>
```
In main view as well:
https://github.com/odoo/odoo/blob/12.0/addons/account/views/account_invoice_view.xml#L511

pedrobaeza committed:
[IMP] account_invoice_inter_company: Move analytic info between invoices

If suitable (being non company specific data), analytic information (account and
tags) is moved from source to destination invoice.

chienandalu commented on account_invoice_inter_company/views/account_invoice_views.xml:
```diff
@@ -0,0 +1,22 @@
+        <field name="inherit_id" ref="account.invoice_supplier_form"/>
+        <field name="mode">primary</field>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='invoice_line_ids']/tree/company_id" position="attributes">
```
I guess, isn't it?
```suggestion
            <xpath expr="//field[@name='invoice_line_ids']/tree/field[@name='company_id']" position="attributes">
```

chienandalu commented on account_invoice_inter_company/views/account_invoice_views.xml:
```diff
@@ -0,0 +1,22 @@
+            <xpath expr="//field[@name='invoice_line_ids']/tree/company_id" position="attributes">
+                <attribute name="readonly">0</attribute>
+            </xpath>
+            <xpath expr="//field[@name='invoice_line_ids']/tree/currency_id" position="attributes">
```
And the same here
```suggestion
            <xpath expr="//field[@name='invoice_line_ids']/tree/field[@name='currency_id']" position="attributes">
```

pedrobaeza commented:
/ocabot merge major

pedrobaeza commented:
/ocabot merge major

pedrobaeza committed:
[FIX] account_invoice_inter_company: Handle invoices in multi-line

Do creation in batch for increasing performance and avoid code to fail in case of
existing invoice line.

pedrobaeza commented:
/ocabot merge major

pedrobaeza committed:
[IMP] account_invoice_inter_company: Isolate invoice generation in tests

Avoid bad interactions with the tests of other modules preventing the generation
of the inter company invoices in that cases.

pedrobaeza commented:
Finally green, so let's merge and forward-port it:

/ocabot merge major
